### ISSUE TITLE
MMU/MMIO performance improvements for sequencing

### ISSUE THREAD

**Issue by OlivierBBB** [Consensys/besu-sequencer-plugins#1440]
# Main point

The contents of (the EVM's) memory are irrelevant for line counting in the `MMU`/`MMIO`. Everything that matters are the underlying (`MMU`) instruction, various offsets and sizes. There is therefore no reason to perform a snapshot of (the EVM's) memory with every instruction when tracing for the purpose of line counting / sequencing.

**Note.** Auxiliary information (such as "does the current instruction justify an `invalidCodePrefixException`) are made ahead of time and don't need to be remembered on top of the sequence of `MMU` operations. Even the information as to whether a `LOG` operation will be carried out will be known when we get to line counting.

# Consequences

To optimize (tracer) memory consumption we may therefore swap the "full (EVM) memory copy" function with a dummy "full copy of memory function" in the sequencer that does nothing. This could be a configuration parameter. This would in theory slash (tracer) memory consumption dramatically for sequencing.

**Comment by letypequividelespoubelles** [comment]
This not true only for the MMU/MMIO, but for a lot of modules. We could have two zkTracer, one "simplified" that does only line counting, and one that does the tracer. We could simplify a lot computations and what we remember in the HUB, MMU, RLPTXN, the three LOG module, all PRC_DATA module, etc

**Comment by ahamlat** [comment]
My understanding of the current implementation is that each module has methods for line counting, like `tracePreOpcode` and `call` methods, and one other method `commit` related to generating the traces, correct ?
I guess we want to move all the extra data methods not needed for line counting to the commit method.

**Comment by letypequividelespoubelles** [comment]
Not exactly. Each module has method to fill it's set of `operations`  : mostly through Besu's hook like tracePreOpcode, traceContextEnter, etc or via direct call (like `callLt` for `WCP` for exemple). In the set of operations we store all informations that are required for the tracing (like full snapshot of memory for MMU/ROM, transaction call data for RLP_TXN, every touched stack item by an opcode for all the stack rows of the HUB, etc...

When we call `lineCount` we get the sum of the lineCount of each `moduleOperations` (it's a little bit smarter as we don't recompute everything from scratch, we use a cache for committed moduleOperation, but the idea is the same).

At the end of the conflation, the `commit` function does, for every module, some needed computation that are not required for the line counting, and perform the tracing (ie fill the .lt file).

So what we could do is having a light zktracer that stores information that are required only for the lineCounting. For a lot of module, it would reduce a lot the memory footprint (ex, for the hub, we could keep the nb of line of each section instead of the full section itself ...

For some module, it wouldn't change anything, as we have to keep in memory the underlying data of the module operation to not having duplicate (eg WCP, EUC, ADD, BIN, SHF, MOD, OOB, MUL (maybe few others). But for all other modules, we coud have way less stored information if we want to perform only the line counting. 

Let's keep in mind that it won't be trivial if we want to do it in a smart and performant way (especially for the HUB). Testing would be easy though, we just would have to test that the light and full zkTracer give the same lineCount for all the test we have.

**Comment by OlivierBBB** [comment]
I whole heartedly concur on the topic of `LOG_DATA` and the precompile modules (`ECDATA`, `BLKMDX`, `SHAKIRA`.)

**Comment by OlivierBBB** [comment]
Since this approach will be applicable to several modules we should have a unified approach, i.e. something like a special subclass of `Module` or so à la `VacuousTracingModule` or so.

**Comment by letypequividelespoubelles** [comment]
I would be in favour of having sequencerModule when the behaviour is too different (HUB, etc). We have to see what can (should ?) be factorised between the tracing and sequencing module. 

We could have, for example an interface HUB , which is implemented in a sequencerHub and tracingHub.

**Comment by OlivierBBB** [comment]
I'm not convinced about the reduction in the HUB ... I mean it's likely possible. But if we do it for the HUB ... then we can go crazy and ... completely forget about the tracing part and just do the line counting. And this could be true in general for most modules.

**Comment by letypequividelespoubelles** [comment]
We now have a `ZkCounter`only :  https://github.com/Consensys/linea-tracer/pull/2034
